### PR TITLE
hyprlandPlugins.hy3: init at 0.32.0

### PR DESCRIPTION
**Pull request by aacebedo**
## Description of changes

Add hyprland plugin support with the first being hy3

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aacebedo** [comment]
@fufexan can you have a look please?

**Comment by fufexan** [comment]
I've thought of a more streamlined approach, you can find it [here](https://github.com/hyprwm/hyprland-plugins/commit/173021806369d4b4d6f99259927c241bb131e0cd). Still up for debate.
@outfoxxed thoughts?

**Comment by outfoxxed** [comment]
@fufexan Imo @aacebedo's approach is more flexible with regard to build systems and unlisted options, but they're extremely similar.

**Comment by aacebedo** [comment]
> I've thought of a more streamlined approach, you can find it [here](https://github.com/hyprwm/hyprland-plugins/commit/173021806369d4b4d6f99259927c241bb131e0cd). Still up for debate. @outfoxxed thoughts?

Why not but I was expecting to have the plugins available inside nixpkgs as hyprland is. I need this because I prefer to have the modules in my flake coming from nixpkgs in priority. I took inspiration from other applications featuring plugins and already present in nixpkgs.

**Review by outfoxxed — COMMENTED** [review]
This seems like a really odd style choice, not sure if its a nixpkgs convention I'm missing though.

**Comment by outfoxxed** [comment]
Maybe `meta` should also have some general information about how to use plugins appended for every plugin

**Comment by aacebedo** [comment]
ES plugins are written this way (https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/servers/search/elasticsearch/plugins.nix) 
vimplugins as kind of similar approach (https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/editors/vim/plugins/generated.nix)

**Comment by outfoxxed** [comment]
I was just commenting on moving `...` to its own line instead of leaving it inline

**Comment by aacebedo** [comment]
> I was just commenting on moving `...` to its own line instead of leaving it inline

ok. I agree this can be changed.

**Comment by aacebedo** [comment]
> Maybe `meta` should also have some general information about how to use plugins appended for every plugin

You mean in the description field or as comment in the source? Usually I take the description from upstream in order the package is as faithful as possible to the intent of the author.

**Comment by outfoxxed** [comment]
Yeah I meant append to it with some general instructions on how to install a plugin, since you can't just put them in path.

**Comment by aacebedo** [comment]
> Yeah I meant append to it with some general instructions on how to install a plugin, since you can't just put them in path.

ok noted

**Comment by aacebedo** [comment]
done

**Comment by fufexan** [comment]
It would be nice to expose the `mkHyprlandPlugin` function, as it can be used for declaring plugins outside nixpkgs.

**Comment by outfoxxed** [comment]
> done

I meant generically for all plugins via the builder.
It probably makes sense to say something like "Plugins can be installed via a [`plugin` entry in the hyprland config](https://wiki.hyprland.org/Plugins/Using-Plugins/#installing--using-plugins) or home manager.

> It would be nice to expose the mkHyprlandPlugin function, as it can be used for declaring plugins outside nixpkgs.

yeah this

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (fufexan) [inline pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:7]:
```suggestion
{ lib
, fetchFromGitHub
, pkg-config
, gcc13Stdenv
, hyprland
, callPackage
}:
```

How about this? This way each plugin can specify its dependencies like a normal package would.

Comment (aacebedo) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:7]: Not sure to follow you on this. You would create one file for each plugin then and add their dependencies?

**Comment by aacebedo** [comment]
> > done
> 
> I meant generically for all plugins via the builder. It probably makes sense to say something like "Plugins can be installed via a [`plugin` entry in the hyprland config](https://wiki.hyprland.org/Plugins/Using-Plugins/#installing--using-plugins) or home manager.
> 
> > It would be nice to expose the mkHyprlandPlugin function, as it can be used for declaring plugins outside nixpkgs.
> 
> yeah this

changed for the description

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (fufexan) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:7]: Well, you'd make a plugin like this, in the same file:
```nix
hyprbars = callPackage (
{ cmake
, pixman
, cairo}:
mkHyprlandPlugin {
  pname = "hyprbars";
  version = "0.1";

  src = fetchFromGitHub ...;

  buildInputs = [
    cairo
    pixman
  ];

  nativeBuildInputs = [cmake];

  # ...
}) {};
```
Sure, it looks a little weird, but we can integrate the `callPackage` call inside `mkHyprlandPlugin`.

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (outfoxxed) [inline pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:22]: There is no nixos option for this because it can only be specified inside the config. I'd also write it as `\n\n<text?`

Comment (aacebedo) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix]: my bad I was convinced it was also existing. Corrected

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (aacebedo) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:7]: Ok I fixed it. @fufexan can you have a look please?

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (fufexan) [inline pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:18]:
```suggestion
          "\n\nPlugins can be installed via a plugin entry in the Hyprland NixOS or Home Manager options.";
```

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (fufexan) [inline pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:35]: Does this not conflict with Hyprland's `meson` and `ninja`?

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (aacebedo) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:18]: done

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (aacebedo) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:35]: I don't think so. Especially because I removed the concatenation of nativeBuildInputs of hyprland in the function.

Comment (fufexan) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:35]: Ah, makes sense then.

**Review by fufexan — APPROVED** [review]

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:35]: Shouldn't pkg-config be moved in the wrapper above to match hyperland?

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:10]:
```suggestion
    args@{ pluginName, ... }:
    gcc13Stdenv.mkDerivation (args // {
```
thats the standard reserved variable for those kinds of things

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:18]:
```suggestion
      pname = pluginName;

      nativeBuildInputs = args.nativeBuildInputs or [ ];
      buildInputs = [ hyprland ]
        ++ hyprland.buildInputs
        ++ (a.buildInputs or [ ]);

      meta = args.meta // {
        lognDescription = ''
          Plugins can be installed via a plugin entry in the Hyprland NixOS or Home Manager options.
        '';
```

The amendment to description should go to longDescription because it is to long.

**Review by SuperSandro2000 — COMMENTED** [review]
I am not sure why ofborg fails. In question try rebasing but this should be fixed before merging.

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (aacebedo) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix]: fixed

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (aacebedo) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix]: you were right fixed

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (aacebedo) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix]: fixed

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (aacebedo) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix]: fixed

**Comment by aacebedo** [comment]
> I am not sure why ofborg fails. In question try rebasing but this should be fixed before merging.

ofborg still fails :(

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:5697]:
```suggestion
  hyprlandPlugins = recurseIntoAttrs (callPackage ../applications/window-managers/hyprwm/hyprland/plugins.nix { });
```

see https://gist.github.com/GrahamcOfBorg/f430b5f70beffcd850e48764cd2efbad#file-output-path-comparison-L141

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (aacebedo) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix]: according to ofborg checks longdescription does not exist

**Comment by aacebedo** [comment]
@SuperSandro2000 problems fixed can you merge please ?

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:32]:
```suggestion
          rev = "hl${version}";
```

**Review by SuperSandro2000 — APPROVED** [review]
sorry, can't merge it :(

File: pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix:25]:
```suggestion
      mkHyprlandPlugin rec {
```

Comment (aacebedo) [reply pkgs/applications/window-managers/hyprwm/hyprland/plugins.nix]: It is fixed

**Comment by aacebedo** [comment]
@SuperSandro2000 one last look hopefully :)

**Pull request merged**
